File: springframework/7.0.8/analyzed/spring-context/src/main/java/org/springframework/scheduling/config/Task.java
/*
 * Copyright 2002-present the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.scheduling.config;

import java.time.Instant;

import org.jspecify.annotations.Nullable;

import org.springframework.scheduling.SchedulingAwareRunnable;
import org.springframework.util.Assert;

/**
 * 定义作为任务执行的 {@code Runnable} 的持有者类，
 * 通常在计划时间或间隔执行。各种调度方式见子类层次结构。
 *
 * @author Chris Beams
 * @author Juergen Hoeller
 * @author Brian Clozel
 * @since 3.2
 */
public class Task {

	private final Runnable runnable;

	private TaskExecutionOutcome lastExecutionOutcome;


	/**
	 * 创建新的 {@code Task}。
	 * @param runnable 要执行的底层任务
	 */
	public Task(Runnable runnable) {
		Assert.notNull(runnable, "Runnable must not be null");
		this.runnable = new OutcomeTrackingRunnable(runnable);
		this.lastExecutionOutcome = TaskExecutionOutcome.create();
	}


	/**
	 * 返回执行底层任务的 {@link Runnable}。
	 * <p>注意，未必返回 {@link Task#Task(Runnable) 原始 runnable}，
	 * 框架可能为其包装以提供额外支持。
	 */
	public Runnable getRunnable() {
		return this.runnable;
	}

	/**
	 * 返回上次任务执行的结果。
	 * @since 6.2
	 */
	public TaskExecutionOutcome getLastExecutionOutcome() {
		return this.lastExecutionOutcome;
	}

	@Override
	public String toString() {
		return this.runnable.toString();
	}


	private class OutcomeTrackingRunnable implements SchedulingAwareRunnable {

		private final Runnable runnable;

		public OutcomeTrackingRunnable(Runnable runnable) {
			this.runnable = runnable;
		}

		@Override
		public void run() {
			try {
				Task.this.lastExecutionOutcome = Task.this.lastExecutionOutcome.start(Instant.now());
				this.runnable.run();
				Task.this.lastExecutionOutcome = Task.this.lastExecutionOutcome.success();
			}
			catch (Throwable exc) {
				Task.this.lastExecutionOutcome = Task.this.lastExecutionOutcome.failure(exc);
				throw exc;
			}
		}

		@Override
		public boolean isLongLived() {
			if (this.runnable instanceof SchedulingAwareRunnable sar) {
				return sar.isLongLived();
			}
			return SchedulingAwareRunnable.super.isLongLived();
		}

		@Override
		public @Nullable String getQualifier() {
			if (this.runnable instanceof SchedulingAwareRunnable sar) {
				return sar.getQualifier();
			}
			return SchedulingAwareRunnable.super.getQualifier();
		}

		@Override
		public String toString() {
			return this.runnable.toString();
		}
	}

}

File: ragflow/0.26.4/analyzed/internal/entity/task.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

// task.go — 文档解析任务实体：记录单文档分页切片、进度、重试与 chunk 产出摘要。
//

package entity

import "time"

// Task 文档处理任务 GORM 实体（表 task）
type Task struct {
	// ID 任务主键
	ID              string     `gorm:"column:id;primaryKey;size:32" json:"id"`
	// DocID 关联文档 ID
	DocID           string     `gorm:"column:doc_id;size:32;not null;index" json:"doc_id"`
	// FromPage 处理起始页码（PDF 等分页格式）
	FromPage        int64      `gorm:"column:from_page;default:0" json:"from_page"`
	// ToPage 处理结束页码上界
	ToPage          int64      `gorm:"column:to_page;default:100000000" json:"to_page"`
	// TaskType 任务类型（parse/embed 等）
	TaskType        string     `gorm:"column:task_type;size:32;not null;default:''" json:"task_type"`
	// Priority 调度优先级，数值越大越优先
	Priority        int64      `gorm:"column:priority;default:0" json:"priority"`
	// BeginAt 任务开始执行时间
	BeginAt         *time.Time `gorm:"column:begin_at;index" json:"begin_at,omitempty"`
	// ProcessDuration 处理耗时（秒）
	ProcessDuration float64    `gorm:"column:process_duration;default:0" json:"process_duration"`
	// Progress 进度 0~1
	Progress        float64    `gorm:"column:progress;default:0;index" json:"progress"`
	// ProgressMsg 进度详情或错误说明
	ProgressMsg     *string    `gorm:"column:progress_msg;type:longtext" json:"progress_msg,omitempty"`
	// RetryCount 已重试次数
	RetryCount      int64      `gorm:"column:retry_count;default:0" json:"retry_count"`
	// Digest 任务结果摘要（如 token 统计）
	Digest          *string    `gorm:"column:digest;type:longtext" json:"digest,omitempty"`
	// ChunkIDs 产出 chunk ID 列表（JSON/逗号分隔）
	ChunkIDs        *string    `gorm:"column:chunk_ids;type:longtext" json:"chunk_ids,omitempty"`
	BaseModel
}

// TableName 返回 GORM 表名 task
func (Task) TableName() string {
	return "task"
}

// Worker 轮询 progress<1 的任务；大文档按 from_page/to_page 拆分子任务。retry_count 达上限后标记失败并写入 progress_msg。
